Exercise 1.1

HussainGynai opened the pull request:
Files for exercise 1.1

## What we want to test:
- Knowledge of basic C++ syntax
- Variable declaration 
- How and when to use While and for loops
- How to index into and iterate through vectors
- How to use conditionals
## Exercise 1.1: command-line “move robot”
- Given robot with 4 wheel speed
- Vector of floats
- Continuously take in a character that tells you what direction you turn
- Use while loop to continuously get input 'wasd’ and ‘e’ for end
- If it's not any of those ask for direction again
- Use if else statements to check input
- Increment/decrement the value of each wheel speed by indexing into array
- Use for loop to print out wheel speeds

JasonGibson274 committed:
first pass at an example project 1

HussainGynai committed:
Created blank, answer and markdown file for exercise 1.1

barulicm commented on assignments/week_1/exercise_1_1/exercise_1_1.md:
```diff
@@ -0,0 +1,161 @@
+# Exercise 1.1
+Welcome to your first exercise of RoboJackets software training! Exercises are shorter an simpler assignments where you will be walked through how to write certain code. This exercise will be testing your knowledge of basic C++ syntax as well as basic programming concepts. For this exercise we expect you to have watched or know the content of the Codecademy C++ course up through vectors, as well  ...
```
```suggestion
Welcome to your first exercise of RoboJackets software training! Exercises are shorter and simpler assignments where you will be walked through how to write certain code. This exercise will be testing your knowledge of basic C++ syntax as well as basic programming concepts. For this exercise we expect you to have watched or know the content of the Codecademy C++ course up through vectors.
```

barulicm commented on assignments/week_1/exercise_1_1/exercise_1_1.md:
```diff
@@ -0,0 +1,161 @@
+
+# Exercise Objective
+In this exercise you will be controlling the movements of a robot with 4 wheels via the command line. The goal is to create a system that will move a robot left, right, forwards or backwards depending on input by the user. Every time the direction of the robot is inputted by the user, the robot should print out the speed each wheel is going. The robot should also continuously check for user input ...
+
```
This section should include a description of how the pretend robot is configured. Is it a tank drive? Omni directional? Be sure to also include a brief explanation of how the robot achieves different directional movement. For example, most of the students are not going to have any familiarity with how omni directional drive bases work, so that would need a good explanation (probably with a picture or two) of how different wheel speeds result in different directions of motion. Be sure to include which wheel index matches which wheel (front, left, etc. or whatever names make sense)

barulicm commented on assignments/week_1/exercise_1_1/exercise_1_1_answer.cpp:
```diff
@@ -0,0 +1,62 @@
+    std::cout << "Use the wasd keys and enter to input a direction, or the E key to stop" << std::endl;
+
+    // Write your code for step 1 here
+    std::vector<float> wheel_speeds(4);
```
Should we be using `std::array` here instead of `std::vector`? Might be a good question to discuss as a group.

barulicm replied:
nvm. they don't know what std::array is yet

barulicm commented on assignments/week_1/exercise_1_1/exercise_1_1_answer.cpp:
```diff
@@ -0,0 +1,62 @@
+            }
+        }
+        else if (direction == 'a') {
+            wheel_speeds[0] -= 0.5;
```
I don't understand why all of these direction changes are adding and subtracting. Shouldn't the wheel speeds be fixed values based on which direction we're trying to go?

HussainGynai replied:
I had them as fixed values before, but in one of the meetings we decided to change it to incrementing/decrementing. I can change it back if we want.

barulicm commented on assignments/week_1/exercise_1_1/exercise_1_1_answer.cpp:
```diff
@@ -0,0 +1,62 @@
+        std::cin >> direction; //At the begining of every iteration look for an input from the user
+        // Write rest of code here (within while loop)
+        if (direction == 'w') {
+            for (int i = 0; i < 4; i++) {
```
We should replace at least one of the for loops in this example with a range-based for loop to give them practice with that. Bonus points for including a discussion on when we would use each type of for loop, but that can also be part of the presentation for this week.

HussainGynai replied:
Wouldn't that mean having to use the & operator since the range based loop is pass by value? I can include it if we want, but new members might get confused. I could use it for the last one where they just print out the wheel speeds but then they don't get the 'number' of the wheel so easily.

kylestach replied:
Yeah, in this particular example it would. @barulicm this is also the solution, so we changing things here doesn't really give them practice with that.

barulicm replied:
The loop below that prints out the wheel speeds could be done without references, so that would be a good one to choose. You can encourage them to use the range-based loop style in the instructions.

HussainGynai replied:
The thing is it then it becomes difficult to print out the index of the wheel with the wheel speed in a nice way, which I think is probably a useful way for them to see if their code works.

kylestach replied:
You could just as easily ask them to print it out as a vector, like `{w0, w1, w2, w3}`. This would allow them to use range-based for.

barulicm replied:
@HussainGynai that's a good point about printing out the index. I missed that it was being used that way. Alright, maybe just forget about that for this exercise. I'll keep an eye out for opportunities to put them in other exercises.

barulicm commented on assignments/week_1/exercise_1_1/exercise_1_1_answer.cpp:
```diff
@@ -0,0 +1,62 @@
+//Do not edit the next 5 lines
+#include <stdlib.h>
```
Does this code actually use something from `stdlib.h`?

barulicm replied:
Can I get an answer on this or can this include line be removed?

barulicm commented on assignments/week_1/exercise_1_1/exercise_1_1_answer.cpp:
```diff
@@ -0,0 +1,62 @@
+            keep_going = false;
+        }
+        else {
+            std::cout << "Sorry I didn't get that, try again" << std::endl;
```
```suggestion
            std::cout << "Sorry, I didn't get that. Try again." << std::endl;
```

barulicm commented on assignments/week_1/exercise_1_1/exercise_1_1_answer.cpp:
```diff
@@ -0,0 +1,62 @@
+            for (int i = 0; i < 4; i++) {
+                wheel_speeds[i] = 0;
+            }
+            std::cout << "Shutting down\n";
```
Per our FAQ
```suggestion
            std::cout << "Shutting down" << std::endl;
```

barulicm commented on assignments/week_1/exercise_1_1/exercise_1_1_answer.cpp:
```diff
@@ -0,0 +1,62 @@
+    wheel_speeds[3] = 0;
+    // Do not edit the next 5 lines
+    bool keep_going = true;
+    char direction;
```
I know we're confident in this example that `direction` is assigned to before we use it, but I feel like we should initialize it to help enforce that habit.
```suggestion
    char direction = 0;
```

barulicm reviewed (changes requested):
Nice start on the write up for this! We'll probably look over this as a group at the meeting since it's our first example.

HussainGynai committed:
Update assignments/week_1/exercise_1_1/exercise_1_1.md

Co-authored-by: Matthew Barulic <matthew.barulic@gmail.com>

HussainGynai committed:
Update assignments/week_1/exercise_1_1/exercise_1_1_answer.cpp

Co-authored-by: Matthew Barulic <matthew.barulic@gmail.com>

HussainGynai committed:
Update assignments/week_1/exercise_1_1/exercise_1_1_answer.cpp

Co-authored-by: Matthew Barulic <matthew.barulic@gmail.com>

HussainGynai committed:
Update assignments/week_1/exercise_1_1/exercise_1_1_answer.cpp

Co-authored-by: Matthew Barulic <matthew.barulic@gmail.com>

JasonGibson274 commented on assignments/week_1/exercise_1_1/exercise_1_1.md:
```diff
@@ -0,0 +1,161 @@
+# Exercise 1.1
+Welcome to your first exercise of RoboJackets software training! Exercises are shorter and simpler assignments where you will be walked through how to write certain code. This exercise will be testing your knowledge of basic C++ syntax as well as basic programming concepts. For this exercise we expect you to have watched or know the content of the Codecademy C++ course up through vectors.
+
+# Exercise Objective
```
Can we include a picture of an example robot with a graphic. I am having trouble getting an idea of what the robot configuration is.

HussainGynai committed:
Update exercise_1_1.md

kylestach commented on assignments/week_1/exercise_1_1/exercise_1_1.md:
```diff
@@ -0,0 +1,163 @@
+# Exercise Objective
+In this exercise you will be controlling the movements of a car like robot with 4 wheels via the command line. The robot looks like a simple car with 4 wheels, each numbered as shown below. All 4 wheel speeds are controlled independently of each other and can move either forward (positive speed) or backward (negative speed). Since the wheels do not pivot, to turn left the left wheels must be movi ...
+
+![alt text](https://github.com/HussainGynai/software-training/blob/master/exercise_1.1_robot.PNG)
```
```suggestion
![alt text](exercise_1.1_robot.PNG)
```

HussainGynai committed:
Update assignments/week_1/exercise_1_1/exercise_1_1.md

Co-authored-by: Kyle Stachowicz <kylestach99@gmail.com>

HussainGynai committed:
Update exercise_1_1_answer.cpp

Changed so the wheel speeds are fixed speeds instead of being incremented/decremented

HussainGynai committed:
Update exercise_1_1.cpp

HussainGynai committed:
Update exercise_1_1.md

Changed instructions to match the fact we are no longer incrementing/decrementing wheel speeds.

HussainGynai committed:
Added robot image

barulicm commented on assignments/week_1/exercise_1_1/exercise_1_1.md:
```diff
@@ -0,0 +1,161 @@
+  for (int i = 0; i < 4; i++) {
+    wheel_speeds[i] = 0;
+  }
+  std::cout << "Shutting down\n";
```
Per our FAQ:
```suggestion
  std::cout << "Shutting down" << std::endl;
```

barulicm commented on assignments/week_1/exercise_1_1/exercise_1_1_answer.cpp:
```diff
@@ -0,0 +1,62 @@
+            }
+        }
+        else if (direction == 'a') {
+            wheel_speeds[0] -= 0.5;
```
This solution code no longer matches the instructions.

barulicm reviewed (changes requested):
This is great and really close to done. There are still some comments from my previous review that have not been addressed though.

barulicm commented:
Ugh, github was not showing me the complete changes while I was reviewing that a moment ago. Please disregard anything that doesn't make sense. I'm going to do a pass now over the actual current change set.

barulicm commented on assignments/week_1/exercise_1_1/exercise_1_1.md:
````diff
@@ -0,0 +1,163 @@
+
+```c++
+//Do not edit the next 5 lines
+#include <stdlib.h>
````
This line was removed from the actual code, but left in the instructions.
```suggestion
```

barulicm commented on assignments/week_1/exercise_1_1/exercise_1_1.md:
```diff
@@ -0,0 +1,163 @@
+#include <vector>
+
+int main(int argc, char** argv) {
+	std::cout << "Use the wasd keys and enter to input a direction, or the E key to stop" << std::endl;
```
nit: replace tab with spaces
```suggestion
    std::cout << "Use the wasd keys and enter to input a direction, or the E key to stop" << std::endl;
```

barulicm commented on assignments/week_1/exercise_1_1/exercise_1_1.md:
```diff
@@ -0,0 +1,163 @@
+#include <iostream>
+#include <vector>
+
+int main(int argc, char** argv) {
```
Omit `argc` and `argv` if we're not using them.
```suggestion
int main() {
```

barulicm commented on assignments/week_1/exercise_1_1/exercise_1_1.md:
```diff
@@ -0,0 +1,163 @@
+   
+    // Do not edit the next 5 lines
+    bool keep_going = true;
+    char direction;
```
You fixed this in the solution, but the starter code and instructions need to be updated to initialize this variable.
```suggestion
    char direction = 0;
```

barulicm commented on assignments/week_1/exercise_1_1/exercise_1_1.md:
````diff
@@ -0,0 +1,163 @@
+  <summary>Answer</summary>
+
+  ```c++
+  std::cout << "Sorry I didn't get that, try again" << std::endl;
````
This should be updated to match the solution code.
```suggestion
9. Within the else statement create a print statement which prints “Sorry, I didn’t get that. Try again.” on a line by itself.
<details>
  <summary>Answer</summary>

  ```c++
  std::cout << "Sorry, I didn't get that.  Try again." << std::endl;
```

barulicm reviewed (changes requested):
Ok, reviewing all of the changes, for real this time. The main thing standing in the way of approval is just re-synchronizing the files so they're consistent with each other after the changes you've made during review.

HussainGynai committed:
Update exercise_1_1.md

HussainGynai committed:
Update assignments/week_1/exercise_1_1/exercise_1_1.md

Co-authored-by: Matthew Barulic <matthew.barulic@gmail.com>

HussainGynai committed:
Update assignments/week_1/exercise_1_1/exercise_1_1.md

Co-authored-by: Matthew Barulic <matthew.barulic@gmail.com>

HussainGynai committed:
Update assignments/week_1/exercise_1_1/exercise_1_1.md

Co-authored-by: Matthew Barulic <matthew.barulic@gmail.com>

HussainGynai committed:
Update assignments/week_1/exercise_1_1/exercise_1_1.md

Co-authored-by: Matthew Barulic <matthew.barulic@gmail.com>

HussainGynai committed:
Update assignments/week_1/exercise_1_1/exercise_1_1.md

Co-authored-by: Matthew Barulic <matthew.barulic@gmail.com>

HussainGynai committed:
Update exercise_1_1.cpp

barulicm reviewed (approved):
Thanks!

barulicm committed:
Merge remote-tracking branch 'origin/master' into exercise_1_1
